csrc/preseg_passes/segment_inplace_update.cpp

```diff
@@ -0,0 +1,156 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <deque>
+#include <vector>
+
+#include <fusion.h>
+#include <id_model/id_model.h>
+#include <id_model/to_string.h>
+#include <ir/utils.h>
+#include <ops/alias.h>
+#include <preseg_passes/segment_inplace_update.h>
+
+namespace nvfuser::preseg_passes {
+// When an intermediate tensorview is aliased to a fusion input,
+// a RW race occurs, when the intermediate tensorview or
+// the aliased input is in path of a broadcast.
+// This preseg pass :
+// 1. Finds any tensorviews used in inplace updates (AllocationType:ReuseBuffer)
+// in the fusion
+// 2. Traverses the fusion graph starting from broadcast ops and stores all
+// direct/indirect producer/consumer tensorviews.
+// 3. For all aliased tensorviews, if the aliased tensorview or the aliased
+// input is present in the set of visited tensorviews in step 2, we insert a
+// segment set and set to force a separate copy kernel. Additionally,
+// we check for implict broadcasts if any aliased input already has a
+// broadcast dimension that is concretized later in the fusion. This ensures
+// that all write operations to the fusion inputs occur after the read
+// operations have completed. See Issue #2664: https://
+// github.com/NVIDIA/Fuser/issues/2664
+namespace {
+void insertSegmentSet(Fusion* fusion) {
+  std::vector<TensorView*> aliased_tvs;
+  std::vector<BroadcastOp*> bcast;
+
+  // Find all tensorviews which are used in inplace updates
+  for (TensorView* tv : fusion->allTvs()) {
+    if (fusion->getOutputAlias(tv).type == AllocationType::ReuseBuffer) {
+      aliased_tvs.push_back(tv);
+    }
+  }
+
+  // Return early if there is no inplace update
+  if (aliased_tvs.empty()) {
+    return;
+  }
+
+  // fusion->exprs() is a topologically sorted list. Filter out the broadcast
+  // ops from the list.
+  auto all_exprs = fusion->exprs();
+  auto all_bcast_ops = ir_utils::filterByType<BroadcastOp>(all_exprs);
+
+  // Traverse and store all direct/indirect consumer tensorviews of these
+  // broadcast nodes If the tensorview has been visited, return --> this means
+  // that we have already traversed that branch
+  std::unordered_set<TensorView*> visited_tvs;
```

Comment:
There are multiple ways to do this traversal using the IdModel approach of checking for concretized broadcasts starting from aliased inputs too:
Let number of number of tensorviews = M, aliased_inputs = N in the given fusion.
- For each aliased input, traverse the graph and if any reachable tensorview has a concretized broadcast, the corresponding aliased output/tv is a candidate for inserting a segment set. This requires O(M) space to store visited tensorviews with concretized bcast and O(N) graph traversals.
- Traverse the graph starting from all aliased inputs and only visit each tensorview once, recording which tensorviews have a concretized bcast. After this, propagate this information upwards towards aliased inputs. This ensures, if during the graph traversal for any aliased input, a node was already visited and hence, the traversal did not go further, we correctly mark that aliased input/output for segment set when a concretized bcast is present further down in that branch. This requires O(M) space, and 2 graph traversals.
- The current approach requires O(M) space to record visited tensorviews, and 1 graph traversal.

This motivated using a hybrid solution which uses:
1. IdModel to check for concretized bcast in aliased inputs (the implicit bcast case)
2. Existing solution of traversing starting from bcast ops, since it required minimum number of graph traversals.

By graph traversal -- I am referring to the number of times each node (tv/expr) is visited.